src/status_im/communities/core.cljs

```diff
@@ -131,18 +131,29 @@
                db
                chat-ids)})
 
-(rf/defn handle-community
-  [{:keys [db]} {:keys [id] :as community}]
-  (when id
-    {:db (assoc-in db [:communities id] (<-rpc community))}))
+(rf/reg-event-fx
+ :communities/handle-community
+ (fn [{:keys [db]}
+      [{token-permissions :tokenPermissions
+        :keys             [token-permissions-check joined id]
+        :as               community}]]
+   (when id
+     (let [has-token-permissions? (not (seq token-permissions))]
+       (cond-> {:db (assoc-in db [:communities id] (<-rpc community))}
```

Comment:
oh i already don't like this approach, it's not about this code, overall

Reply:
fx/defn was implemented to improve such cases

Reply:
`{:fx (map` looks strange, are use sure it will work?

```clojure
{:fx [{:db :fx} {:db :fx}]}
```

Reply:
probably it should be something like
```clojure
(defn handle-communities
  [cofx [communities]]
  (reduce 
(fn [cofx {:keys [id]}]
(let [{:keys [db fx]} (handle-community{:db (:db cofx)} id)]
   {:db db
     :fx (conj (:fx cofx} fx))
 {:db db :fx []} 
communities)})
  ```

Reply:
Your are totally correct here @flexsurfer, my code snippet wasn't tested and it's incorrect, that part about multiple `:db` effects won't work.

In the ideal world (my personal view) db transformations should be decoupled from re-frame concepts of effects. It should be just pure functions that can be composed and if we switch re-frame to something else they would continue to work. I have yet to experiment in a branch with this idea in `status-mobile` to see how well it would work and if it would really make our lives better.

Reply:
To summarize my previous long comment, if there's no measurable/significant problem with the multiple dispatches for each community (because they will be processed in one go due to how re-frame consumes the queue) then I think it's not worth bothering with merging all the effects.

Reply:
re-frame handle events in the current queue non-stop
we're doing the reduce non-stop as well.

I don't know. Seems not much difference to me. Are there any expensive code in re-frame's event queue iterating section?

Reply:
that only happens if you dispatch all events in one js task, but you will have 3 queues , ideally we should never dispatch from dispatch handler

Reply:
> but you will have 3 queues

I think there will be only 2 queues. The running one and the next one.

https://github.com/day8/re-frame/blob/032d93ea55ca19b0a4808a161b09c4e4c39a5b6f/src/re_frame/router.cljc#L36-L41

`rf/merge` make it that each queue will always have only one event to process by merging effects in advance, but that's what re-frame is doing
I think we should consider using `requestAnimationFrame` and `InteractionManager.runAfterInteractions`.

And implement our own `EventQueue` type, like what react 18 does, when we're facing performance problems. https://github.com/facebook/react/blob/main/packages/scheduler/src/forks/Scheduler.js#L270
So that we can at least rely on existing docs from frameworks/tools we are using

Reply:
> And implement our own EventQueue type, like what react 18 does, when we're facing performance problems.

100% 👍🏼 re-frame's queuing algorithm wasn't battle tested for mobile apps. Although I think it's not a problem the vast majority of the time.

The other thing we can do meanwhile because we don't have our own queue implementation (and we don't even know if that could work as well as we want) we can just use `rf/merge`, but I'd only go this route if there's a measured problem, otherwise as you and I @yqrashawn said, but also in other PRs, re-frame consumes scheduled events in one go, so it behaves similarly to `rf/merge` anyway in most occasions.